### PR TITLE
Failover detach

### PR DESCRIPTION
**Pull request by arutk**
Signed-off-by: Adam Rutkowski <adam.j.rutkowski@intel.com>

### CODE REVIEW

File: src/mngt/ocf_mngt_cache.c [hunk 1]
```diff
@@ -2039,65 +2042,78 @@ struct ocf_pipeline_properties _ocf_mngt_cache_bind_pipeline_properties = {
...
+	if (cache_volume->opened) {
+		ret = ocf_volume_init(&context->cache_volume, 0, NULL, false);
+		if (!ret)
+			OCF_PL_FINISH_RET(pipeline, ret);
 
-	ocf_volume_move(&context->cache_volume, &cache->device->volume);
-	context->flags.volume_stored = true;
+		ocf_volume_move(&context->cache_volume, &cache->device->volume);
+		context->flags.volume_stored = true;
+	}
```
Comment (rafalste) [inline src/mngt/ocf_mngt_cache.c:2099]: Same condition checked twice.

File: src/mngt/ocf_mngt_cache.c [hunk 2]
```diff
@@ -2385,22 +2406,82 @@ static void _ocf_mngt_cache_bind(ocf_cache_t cache,
...
-	if (!ocf_cache_is_passive(cache))
+	if (!ocf_cache_is_standby(cache))
 		OCF_CMPL_RET(cache, priv1, priv2, -OCF_ERR_CACHE_EXIST);
 
 	result = ocf_pipeline_create(&pipeline, cache,
 			&_ocf_mngt_cache_activate_pipeline_properties);
 	if (result)
 		OCF_CMPL_RET(cache, priv1, priv2, -OCF_ERR_NO_MEM);
 
+
```
Comment (rafalste) [inline src/mngt/ocf_mngt_cache.c:2484]: double new line

File: inc/ocf_def.h [hunk 3]
```diff
@@ -124,8 +124,8 @@ typedef enum {
 	ocf_cache_state_incomplete = 3, //!< ocf_cache_state_incomplete
 		/*!< OCF cache has at least one inactive core */
 
-	ocf_cache_state_passive = 4,     //!< ocf_cache_state_passive
-		/*!< OCF is currently in passive mode */
+	ocf_cache_state_standby = 4,     //!< ocf_cache_state_standby
+		/*!< OCF is currently in standby mode */
```
[the remark this answers is not in the document]
Comment (arutk) [reply inc/ocf_def.h:128]: As it is possible to activate cache with or without detaching the cache drive, I think it seems more user friendly to call both "standby" - with the cache drive attached or not. I think it makes the flow easier to describe: you can always activate standby cache, regardless of whether the drive is detached.

File: inc/ocf_mngt.h [hunk 4]
```diff
@@ -426,15 +436,15 @@ struct ocf_mngt_cache_device_config {
...
 {
 	cfg->cache_line_size = ocf_cache_line_size_none;
 	cfg->open_cores = true;
 	cfg->force = false;
-	cfg->perform_test = true;
 	cfg->discard_on_start = true;
-	cfg->volume_params = NULL;
+	cfg->device.perform_test = true;
+	cfg->device.volume_params = NULL;
 }
```
[the remark this answers is not in the document]
Comment (arutk) [reply inc/ocf_mngt.h:448]: Do you have any specific arguments against this? From my point of view the new name is more precise (as this (outer) structure does not describe only a device but some management operation params as well).

File: src/mngt/ocf_mngt_cache.c (hunk 1 above)
Comment (arutk) [reply src/mngt/ocf_mngt_cache.c]: all right, fixzed

File: src/mngt/ocf_mngt_cache.c [hunk 5]
```diff
@@ -2409,7 +2490,21 @@ static void _ocf_mngt_cache_activate(ocf_cache_t cache,
...
+	context->cfg.discard_on_start = false;
+
+
+	if (!ocf_refcnt_frozen(&cache->refcnt.metadata)) {
+		ENV_BUG_ON(!cache->device->volume.opened);
+		ocf_refcnt_freeze(&cache->refcnt.metadata);
+		ocf_refcnt_register_zero_cb(&cache->refcnt.metadata,
+			(ocf_refcnt_cb_t)ocf_pipeline_next, pipeline);
+		return;
+	}
```
[the remark this answers is not in the document]
Comment (arutk) [reply src/mngt/ocf_mngt_cache.c:2523]: sure

**Pull request merged**